java-checks/src/main/java/org/sonar/java/checks/SwitchAtLeastThreeCasesCheck.java

```diff
@@ -20,34 +20,40 @@
 package org.sonar.java.checks;
 
 import org.sonar.check.Rule;
-import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
 import org.sonar.plugins.java.api.JavaFileScannerContext;
-import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
 import org.sonar.plugins.java.api.tree.CaseGroupTree;
 import org.sonar.plugins.java.api.tree.SwitchStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.List;
 
 @Rule(key = "S1301")
-public class SwitchAtLeastThreeCasesCheck extends BaseTreeVisitor implements JavaFileScanner {
+public class SwitchAtLeastThreeCasesCheck extends IssuableSubscriptionVisitor {
 
   private JavaFileScannerContext context;
 
   @Override
-  public void scanFile(JavaFileScannerContext context) {
-    this.context = context;
-    scan(context.getTree());
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.SWITCH_STATEMENT);
   }
 
   @Override
-  public void visitSwitchStatement(SwitchStatementTree tree) {
+  public void visitNode(Tree tree) {
+    SwitchStatementTree switchStatementTree = (SwitchStatementTree) tree;
     int count = 0;
-    for (CaseGroupTree caseGroup : tree.cases()) {
+    for (CaseGroupTree caseGroup : switchStatementTree.cases()) {
+      if (caseGroup.labels().stream()
+        .flatMap(label -> label.expressions().stream())
+        .anyMatch(expression -> expression.is(Tree.Kind.TYPE_PATTERN, Tree.Kind.RECORD_PATTERN, Tree.Kind.GUARDED_PATTERN))) {
```

Comment:
I think that to make this method super-simple to read, I would extract this part into a method:

```java
if (containsPatterns(caseGroup)) {
  // usage of pattern improve readability and therefore invalidate use of if/else instead of a switch
  return;
}

// ...

private static boolean containsPatterns(CaseGroupTree caseGroup) {
  return caseGroup.labels().stream()
    .flatMap(label -> label.expressions().stream())
    .anyMatch(expression -> expression.is(Tree.Kind.TYPE_PATTERN, Tree.Kind.RECORD_PATTERN, Tree.Kind.GUARDED_PATTERN));
}
```